### PR TITLE
linuxKernel.kernels: patch out nvme regression on 2017ish Dell laptops

### LINKED ISSUE #253418: Storage issues on XPS 15 9560 due to kernel regression; failing cryptsetup
There are several reports of storage issues all on this one laptop model here, so I am filing a bug: https://discourse.nixos.org/t/nvme-drive-not-detecting-after-calameres-initiates/32108/

Live posts of me doing debugging: https://matrix.to/#/!DBFhtjpqmJNENpLDOv:nixos.org/$9zWPOBSxYIou5aEGILjwnDFgEJLuIi7JTuUUo7iQkyw?via=nixos.org&via=matrix.org&via=tchncs.de

Results of testing on my machine yesterday:
Kernels I believe are ok: 6.4.9, 6.1.44
Kernels I believe are bad: 6.5.0, 6.5.1, 6.1.51, 6.1.49(?)

Debug log of a failing boot: https://gist.github.com/lf-/58fb6bfd13e4f3d09d8e2c39b279b46a

### Describe the bug
Various storage-not-detected/io error symptoms on the XPS 15 9560, with the internal SSD. Seems to not matter *massively* what SSD it is, since there are people having it with the original SSD as far as I can tell.

Reproduced on this aftermarket drive (there's a firmware rev in here somewhere right?):

```
$ nvme id-ctrl /dev/nvme0
<snip>
mn        : WD Blue SN570 1TB
fr        : 234110WD
rab       : 4
ieee      : 001b44
cmic      : 0
mdts      : 7
cntlid    : 0
ver       : 0x10400
rtd3r     : 0x7a120
rtd3e     : 0xf4240
oaes      : 0x200
ctratt    : 0x2
rrls      : 0
cntrltype : 1
<snip>
```

Most relevant part of failed boot log:

```
Sep 04 11:56:38 localhost kernel: nvme nvme0: controller is down; will reset: CSTS=0xffffffff, PCI_STATUS=0xffff
Sep 04 11:56:38 localhost kernel: nvme nvme0: Does your device have a faulty power saving mode enabled?
Sep 04 11:56:38 localhost kernel: nvme nvme0: Try "nvme_core.default_ps_max_latency_us=0 pcie_aspm=off" and report a bug
Sep 04 11:56:38 localhost kernel: nvme 0000:04:00.0: Unable to change power state from D3cold to D0, device inaccessible
Sep 04 11:56:38 localhost kernel: nvme nvme0: Disabling device after reset failure: -19
Sep 04 11:56:38 localhost systemd-cryptsetup[169]: Device /dev/disk/by-uuid/b80aedf8-ddd4-46fa-8d09-5215d5f286b9 READ lock released.
Sep 04 11:56:38 localhost systemd-cryptsetup[169]: IO error while decrypting keyslot.
Sep 04 11:56:38 localhost systemd-cryptsetup[169]: Keyslot 0 (luks2) open failed with -5.
```

I have not yet tried the alleged workaround given in the message here. I might try it on NixCon hacking day when I don't need my computer to work.

### Steps To Reproduce

1. Use bad kernel
2. ???
3. Suffering

### Expected behavior

Disk works, system boots.

### Additional context
This is a regression between nixos-unstable revisions ce5e4a6 and 3efb0f6 which I have debugged to be kernel-version induced. systemd and cryptsetup versions are constant across both

### Notify maintainers
@TredwellGit @Ma27 @NeQuissimus @alyssais @thoughtpolice 

### Metadata

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.9, NixOS, 23.11 (Tapir)
, 23.11.20230902.e569908`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

### PR DESCRIPTION
**Pull request by lf-**
## Description of changes

Fixes #253418

I have personally tested kernel 6.1 and confirmed that my machine now boots
again (yay!). I will set off the rest of them building overnight but may not
get around to actually validating on hardware very soon.

This is the regression this patches around:
https://bugzilla.kernel.org/show_bug.cgi?id=217802

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
Inclined to accept this since OpenSUSE has also reverted, but I'm a bit nervous that this was reported a month ago and still nothing has happened upstream AFAICT?

**Comment by alyssais** [comment]
> > Inclined to accept this since OpenSUSE has also reverted, but I'm a bit nervous that this was reported a month ago and still nothing has happened upstream AFAICT?
> 
> I think that one of their hands isn't talking to the other, probably because lkml is a pain so the users and packagers didn't write in the thread. They could probably use a message reiterating that opensuse reverted downstream, including the bugzilla link, and that it is not just one weird hardware combo and happens to all copies of that laptop model.

How about submitting a revert patch upstream, pointing this out, and linking to the previous LKML and Bugzilla discussion?

**Comment by lf-** [comment]
> > > Inclined to accept this since OpenSUSE has also reverted, but I'm a bit nervous that this was reported a month ago and still nothing has happened upstream AFAICT?
> > 
> > 
> > I think that one of their hands isn't talking to the other, probably because lkml is a pain so the users and packagers didn't write in the thread. They could probably use a message reiterating that opensuse reverted downstream, including the bugzilla link, and that it is not just one weird hardware combo and happens to all copies of that laptop model.
> 
> How about submitting a revert patch upstream, pointing this out, and linking to the previous LKML and Bugzilla discussion?

I think I am hilariously more able to use git-send-email than I am able to practically send replies to the list so I may well do that today.

**Comment by alyssais** [comment]
> I think I am hilariously more able to use git-send-email than I am able to practically send replies to the list so I may well do that today.

(You can also use git-send-email to send non-patch replies)

**Comment by lf-** [comment]
Alright, I have sent off a patch into [this very specific ether](https://lore.kernel.org/lkml/20230918170831.1677690-3-lists@jade.fyi/). You have been cc'd :)

File: pkgs/os-specific/linux/kernel/patches.nix
Comment (alyssais) [inline pkgs/os-specific/linux/kernel/patches.nix:25]: For next time, don't worry about this now unless you want to:

```suggestion
    patch = fetchurl {
```

You're fetching a static file, so there's no need for normalisation and stuff.  (And it's best to avoid that where possible as what "normalised" means depends on a specific version of patchutils.)

Alternatively, if there aren't any conflicts, you could `fetchpatch` the original commit and pass `revert = true;`, which is nice and declarative.

**Review by alyssais — APPROVED** [review]
Let's hope upstream reverts soon.  I'll give my fellow kernel maintainers a short final chance to object.

**Comment by Ma27** [comment]
ftr I meant to leave a comment before merging, but given that OpenSUSE did the same and this fixes an actual case of a machine not booting up, I'm perfectly fine with this.

**Pull request merged**
